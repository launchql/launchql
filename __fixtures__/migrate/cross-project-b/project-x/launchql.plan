%syntax-version=1.0.0
%project=project-x
%uri=https://github.com/test/project-x

core_schema 2024-01-01T00:00:00Z test <test@example.com> # Create core schema
core_types [core_schema] 2024-01-02T00:00:00Z test <test@example.com> # Create core types
@x1.0.0 2024-01-02T12:00:00Z test <test@example.com> # Core system v1.0.0
core_functions [core_types] 2024-01-03T00:00:00Z test <test@example.com> # Core utility functions
@x1.1.0 2024-01-03T12:00:00Z test <test@example.com> # Core system v1.1.0
core_extensions [core_functions] 2024-01-04T00:00:00Z test <test@example.com> # Extended core features
@x2.0.0 2024-01-04T12:00:00Z test <test@example.com> # Core system v2.0.0