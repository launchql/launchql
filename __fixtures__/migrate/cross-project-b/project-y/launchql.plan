%syntax-version=1.0.0
%project=project-y
%uri=https://github.com/test/project-y

# Project Y depends on specific versions of Project X
auth_schema [project-x:@x1.0.0] 2024-01-05T00:00:00Z test <test@example.com> # Auth schema requires core v1.0.0
auth_users [auth_schema project-x:core_types] 2024-01-06T00:00:00Z test <test@example.com> # User tables
@y1.0.0 2024-01-06T12:00:00Z test <test@example.com> # Auth system v1.0.0
auth_sessions [auth_users] 2024-01-07T00:00:00Z test <test@example.com> # Session management
auth_roles [auth_users project-x:@x1.1.0] 2024-01-08T00:00:00Z test <test@example.com> # Roles require core v1.1.0
@y1.1.0 2024-01-08T12:00:00Z test <test@example.com> # Auth system v1.1.0
auth_permissions [auth_roles project-x:core_extensions] 2024-01-09T00:00:00Z test <test@example.com> # Advanced permissions
@y2.0.0 2024-01-09T12:00:00Z test <test@example.com> # Auth system v2.0.0