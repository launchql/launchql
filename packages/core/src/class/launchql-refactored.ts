import { Logger } from '@launchql/logger';
import {
  ConfigManager,
  ModuleManager,
  ExtensionManager,
  PackageManager,
  PlanGenerator
} from '../managers';

export enum ProjectContext {
  Workspace = 'workspace',
  Module = 'module',
  Unknown = 'unknown'
}

/**
 * Refactored LaunchQL class using manager pattern
 * This class serves as a facade to the various managers
 */
export class LaunchQL {
  private logger = new Logger('launchql');
  private configManager: ConfigManager;
  private moduleManager: ModuleManager;
  private extensionManager: ExtensionManager;
  private packageManager: PackageManager;
  private planGenerator: PlanGenerator;

  constructor(cwd: string = process.cwd()) {
    // Initialize managers
    this.configManager = new ConfigManager(cwd);
    this.moduleManager = new ModuleManager(this.configManager);
    this.extensionManager = new ExtensionManager(this.configManager, this.moduleManager);
    this.packageManager = new PackageManager(
      this.configManager,
      this.moduleManager,
      this.extensionManager
    );
    this.planGenerator = new PlanGenerator(
      this.configManager,
      this.moduleManager,
      this.extensionManager
    );
  }

  // Delegate to ConfigManager
  resetCwd(cwd: string): void {
    this.configManager.resetCwd(cwd);
  }

  isInsideAllowedDirs(cwd: string): boolean {
    return this.configManager.isInsideAllowedDirs(cwd);
  }

  ensureModule(): void {
    this.configManager.ensureModule();
  }

  ensureWorkspace(): void {
    this.configManager.ensureWorkspace();
  }

  getContext(): ProjectContext {
    if (this.configManager.isInModule() && this.configManager.isInWorkspace()) {
      return ProjectContext.Module;
    }
    if (this.configManager.isInWorkspace()) return ProjectContext.Workspace;
    return ProjectContext.Unknown;
  }

  isInWorkspace(): boolean {
    return this.configManager.isInWorkspace();
  }

  isInModule(): boolean {
    return this.configManager.isInModule();
  }

  getWorkspacePath(): string | undefined {
    return this.configManager.getWorkspacePath();
  }

  getModulePath(): string | undefined {
    return this.configManager.getModulePath();
  }

  get cwd(): string {
    return this.configManager.getCwd();
  }

  get workspacePath(): string | undefined {
    return this.configManager.getWorkspacePath();
  }

  get modulePath(): string | undefined {
    return this.configManager.getModulePath();
  }

  get config(): any {
    return this.configManager.getConfig();
  }

  get allowedDirs(): string[] {
    return this.configManager.getAllowedDirs();
  }

  // Delegate to ModuleManager
  clearCache(): void {
    this.moduleManager.clearCache();
    this.extensionManager.clearCache();
  }

  getAvailableModules(): LaunchQL[] {
    return this.moduleManager.getAvailableModules();
  }

  getModuleMap(): any {
    return this.moduleManager.getModuleMap();
  }

  getModuleName(): string {
    return this.moduleManager.getModuleName();
  }

  getModuleDependencies(): string[] {
    return this.moduleManager.getModuleDependencies();
  }

  getModuleDependencyChanges(): string[] {
    return this.moduleManager.getModuleDependencyChanges();
  }

  getLatestChange(): string | undefined {
    return this.moduleManager.getLatestChange();
  }

  getLatestChangeAndVersion(): any {
    return this.moduleManager.getLatestChangeAndVersion();
  }

  getModulePlan(): string {
    return this.moduleManager.getModulePlan();
  }

  getModuleSQL(): string[] {
    return this.moduleManager.getModuleSQL();
  }

  getModuleControlFile(): any {
    return this.moduleManager.getModuleControlFile();
  }

  getModuleMakefile(): string | null {
    return this.moduleManager.getModuleMakefile();
  }

  normalizeChangeName(name: string): string {
    return this.moduleManager.normalizeChangeName(name);
  }

  async initModule(name: string): Promise<string> {
    return this.moduleManager.initModule(name);
  }

  // Delegate to ExtensionManager
  getModuleInfo(): any {
    return this.extensionManager.getModuleInfo();
  }

  getModuleExtensions(): string[] {
    return this.extensionManager.getModuleExtensions();
  }

  getAvailableExtensions(): string[] {
    return this.extensionManager.getAvailableExtensions();
  }

  getInstalledExtensions(): string[] {
    return this.extensionManager.getInstalledExtensions();
  }

  async installExtension(packageSpec: string): Promise<void> {
    return this.extensionManager.installExtension(packageSpec);
  }

  async writeExtensions(): Promise<void> {
    return this.extensionManager.writeExtensions();
  }

  // Delegate to PackageManager
  async publishToDist(): Promise<void> {
    return this.packageManager.publishToDist();
  }

  setModuleDependencies(dependencies: string[]): void {
    this.packageManager.setModuleDependencies(dependencies);
  }

  getRequiredModules(): string[] {
    return this.packageManager.getRequiredModules();
  }

  // Delegate to PlanGenerator
  generatePlanFromFiles(): string {
    return this.planGenerator.generatePlanFromFiles();
  }

  writeModulePlan(): void {
    this.planGenerator.writeModulePlan();
  }

  addChangeToProject(name: string, dependencies?: string[]): void {
    this.planGenerator.addChangeToProject(name, dependencies);
  }

  validatePlanConsistency(): { valid: boolean; errors: string[] } {
    return this.planGenerator.validatePlanConsistency();
  }

  // Convenience methods that combine multiple managers
  async install(packages: string[]): Promise<void> {
    for (const pkg of packages) {
      await this.installExtension(pkg);
    }
  }

  async deploy(): Promise<void> {
    // Validate plan first
    const validation = this.validatePlanConsistency();
    if (!validation.valid) {
      throw new Error(`Invalid plan: ${validation.errors.join(', ')}`);
    }

    // Deploy logic would go here
    this.logger.info('Deploy functionality to be implemented');
  }

  async revert(target?: string): Promise<void> {
    // Revert logic would go here
    this.logger.info('Revert functionality to be implemented');
  }

  async verify(): Promise<void> {
    // Verify logic would go here
    this.logger.info('Verify functionality to be implemented');
  }

  async status(): Promise<any> {
    // Status logic would go here
    return {
      workspace: this.getWorkspacePath(),
      module: this.getModulePath(),
      context: this.getContext(),
      modules: this.getAvailableModules().map(m => m.getModuleName()),
      extensions: this.getInstalledExtensions()
    };
  }
}