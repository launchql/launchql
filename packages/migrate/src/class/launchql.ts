import fs, { writeFileSync } from 'fs';
import path from 'path';
import * as glob from 'glob';
import { walkUp } from '../utils';
import { getDeps } from '../deps';
import chalk from 'chalk';

import {
  writeRenderedTemplates,
  moduleTemplate
} from '@launchql/templatizer';

import {
  listModules,
  latestChange,
  latestChangeAndVersion,
  getExtensionsAndModules,
  getExtensionsAndModulesChanges,
  ModuleMap
} from '../modules';

import {
  getExtensionInfo,
  writeExtensions,
  getExtensionName,
  getAvailableExtensions,
  getInstalledExtensions,
  ExtensionInfo,
} from '../extensions';
import { exec, execSync } from 'child_process';

export enum ProjectContext {
  Outside = 'outside',
  Workspace = 'workspace-root',
  Module = 'module',
  ModuleInsideWorkspace = 'module-in-workspace',
}

export interface InitModuleOptions {
  name: string;
  description: string;
  author: string;
  extensions: string[];
}

export class LaunchQLProject {
  public readonly cwd: string;
  public workspacePath?: string;
  public modulePath?: string;
  public config?: any;
  public allowedDirs: string[] = [];

  private _moduleMap?: ModuleMap;
  private _moduleInfo?: ExtensionInfo;

  constructor(cwd: string = process.cwd()) {
    this.cwd = path.resolve(cwd);
    this.workspacePath = this.resolveLaunchqlPath();
    this.modulePath = this.resolveSqitchPath();

    if (this.workspacePath) {
      this.config = this.loadConfig();
      this.allowedDirs = this.loadAllowedDirs();
    }
  }
  
  private resolveLaunchqlPath(): string | undefined {
    try {
      return walkUp(this.cwd, 'launchql.json');
    } catch {
      return undefined;
    }
  }

  private resolveSqitchPath(): string | undefined {
    try {
      return walkUp(this.cwd, 'sqitch.conf');
    } catch {
      return undefined;
    }
  }

  private loadConfig(): any {
    const configPath = path.join(this.workspacePath!, 'launchql.json');
    return JSON.parse(fs.readFileSync(configPath, 'utf8'));
  }

  private loadAllowedDirs(): string[] {
    const globs: string[] = this.config?.packages ?? [];
    const dirs = globs.flatMap(pattern =>
      glob.sync(path.join(this.workspacePath!, pattern))
    );
    return dirs.map(dir => path.resolve(dir));
  }

  isInsideAllowedDirs(cwd: string): boolean {
    return this.allowedDirs.some(dir => cwd.startsWith(dir));
  }

  private createModuleDirectory(modName: string): string {
    this.ensureWorkspace();

    const isRoot = path.resolve(this.workspacePath!) === path.resolve(this.cwd);
    let targetPath: string;

    if (isRoot) {
      const packagesDir = path.join(this.cwd, 'packages');
      fs.mkdirSync(packagesDir, { recursive: true });
      targetPath = path.join(packagesDir, modName);
    } else {

      if (!this.isInsideAllowedDirs(this.cwd)) {
        console.error(chalk.red(`Error: You must be inside one of the workspace packages: ${this.allowedDirs.join(', ')}`));
        process.exit(1);
      }

      targetPath = path.join(this.cwd, modName);
    }

    fs.mkdirSync(targetPath, { recursive: true });
    return targetPath;
  }

  ensureModule(): void {
    if (!this.modulePath) throw new Error('Not inside a module');
  }  

  ensureWorkspace(): void {
    if (!this.workspacePath) throw new Error('Not inside a workspace');
  }  

  getContext(): ProjectContext {
    if (this.modulePath && this.workspacePath) {
      const rel = path.relative(this.workspacePath, this.modulePath);
      const nested = !rel.startsWith('..') && !path.isAbsolute(rel);
      return nested ? ProjectContext.ModuleInsideWorkspace : ProjectContext.Module;
    }

    if (this.modulePath) return ProjectContext.Module;
    if (this.workspacePath) return ProjectContext.Workspace;
    return ProjectContext.Outside;
  }

  isInWorkspace(): boolean {
    return this.getContext() === ProjectContext.Workspace;
  }

  isInModule(): boolean {
    return (
      this.getContext() === ProjectContext.Module ||
      this.getContext() === ProjectContext.ModuleInsideWorkspace
    );
  }

  getWorkspacePath(): string | undefined {
    return this.workspacePath;
  }

  getModulePath(): string | undefined {
    return this.modulePath;
  }

  clearCache() {
    delete this._moduleInfo;
    delete this._moduleMap;
  }

  // ──────────────── Workspace-wide ────────────────

  async getModules(): Promise<LaunchQLProject[]> {
    if (!this.workspacePath || !this.config) return [];

    const dirs = this.loadAllowedDirs();
    const results: LaunchQLProject[] = [];

    for (const dir of dirs) {
      const proj = new LaunchQLProject(dir);
      if (proj.isInModule()) {
        results.push(proj);
      }
    }

    return results;
  }

  getModuleMap(): ModuleMap {
    if (!this.workspacePath) return {};
    if (this._moduleMap) return this._moduleMap;

    this._moduleMap = listModules(this.workspacePath);
    return this._moduleMap;
  }

  getAvailableModules(): string[] {
    const modules = this.getModuleMap();
    return getAvailableExtensions(modules);
  }

  // ──────────────── Module-scoped ────────────────

  getModuleInfo(): ExtensionInfo {
    this.ensureModule();
    if (!this._moduleInfo) {
      this._moduleInfo = getExtensionInfo(this.cwd);
    }
    return this._moduleInfo;
  }

  getModuleName(): string {
    this.ensureModule();
    return getExtensionName(this.cwd);
  }

  getRequiredModules(): string[] {
    this.ensureModule();
    const info = this.getModuleInfo();
    return getInstalledExtensions(info.controlFile);
  }

  setModuleDependencies(modules: string[]): void {
    this.ensureModule();
    writeExtensions(this.cwd, modules);
  }

  private initModuleSqitch(modName: string, targetPath: string): void {
    const cur = process.cwd();
    process.chdir(targetPath);
    execSync(`sqitch init ${modName} --engine pg`, { stdio: 'inherit' });
    const plan = `%syntax-version=1.0.0\n%project=${modName}\n%uri=${modName}`;
    writeFileSync(path.join(targetPath, 'sqitch.plan'), plan);
    process.chdir(cur);
  }

  initModule(options: InitModuleOptions): void {
    this.ensureWorkspace();
    const targetPath = this.createModuleDirectory(options.name);
    writeRenderedTemplates(moduleTemplate, targetPath, options);  
    this.initModuleSqitch(options.name, targetPath);
    writeExtensions(targetPath, options.extensions);
  }

  // ──────────────── Dependency Analysis ────────────────

  getLatestChange(moduleName: string): string {
    const modules = this.getModuleMap();
    return latestChange(moduleName, modules, this.workspacePath!);
  }

  getLatestChangeAndVersion(moduleName: string): { change: string; version: string } {
    const modules = this.getModuleMap();
    return latestChangeAndVersion(moduleName, modules, this.workspacePath!);
  }

  getModuleDependencies(moduleName: string): { native: string[]; modules: string[] } {
    const modules = this.getModuleMap();
    const { native, sqitch } = getExtensionsAndModules(moduleName, modules);
    return { native, modules: sqitch };
  }

  async getModuleDependencyChanges(moduleName: string): Promise<{
    native: string[];
    modules: { name: string; latest: string; version: string }[];
  }> {
    const modules = this.getModuleMap();
    const { native, sqitch } = await getExtensionsAndModulesChanges(moduleName, modules, this.workspacePath!);
    return { native, modules: sqitch };
  }

  // ──────────────── Plans ────────────────

  getModulePlan (): string {
    this.ensureModule();
    const planPath = path.join(this.getModulePath()!, 'sqitch.plan');
    return fs.readFileSync(planPath, 'utf8');
  }

  getModuleControlFile(): string {
    this.ensureModule();
    const info = this.getModuleInfo();
    return fs.readFileSync(info.controlFile, 'utf8');
  }

  getModuleMakefile(): string {
    this.ensureModule();
    const info = this.getModuleInfo();
    return fs.readFileSync(info.Makefile, 'utf8');
  }

  getModuleSQL(): string {
    this.ensureModule();
    const info = this.getModuleInfo();
    return fs.readFileSync(info.sqlFile, 'utf8');
  }

  async generateModulePlan(options: { uri?: string; projects?: boolean }): Promise<string> {
    this.ensureModule();
    const info = this.getModuleInfo();
    const moduleName = info.extname;

    const now =
      process.env.NODE_ENV === 'test'
        ? '2017-08-11T08:11:51Z'
        : new Date().toISOString();

    const planfile: string[] = [
      `%syntax-version=1.0.0`,
      `%project=${moduleName}`,
      `%uri=${options.uri || moduleName}`
    ];

    const { resolved, deps } = await getDeps(this.cwd, moduleName);

    if (options.projects && this.workspacePath) {
      const depData = await this.getModuleDependencyChanges(moduleName);
      const external = depData.modules.map((m) => `${m.name}:${m.latest}`);

      const key = `/deploy/${resolved[0]}.sql`;
      deps[key] ||= [];
      deps[key].push(...external);
    }

    const makeKey = (sqlmod: string) => `/deploy/${sqlmod}.sql`;

    resolved.forEach((res) => {
      if (/:/.test(res)) return;

      const dependencies = deps[makeKey(res)];
      if (dependencies?.length) {
        planfile.push(
          `${res} [${dependencies.join(
            ' '
          )}] ${now} launchql <launchql@5b0c196eeb62> # add ${res}`
        );
      } else {
        planfile.push(`${res} ${now} launchql <launchql@5b0c196eeb62> # add ${res}`);
      }
    });

    return planfile.join('\n');
  }

  async writeModulePlan(
    options: { uri?: string; projects?: boolean }
  ): Promise<void> {
    this.ensureModule();
    const name = this.getModuleName();
    const plan = await this.generateModulePlan(options);
    const moduleMap = this.getModuleMap();
    const mod = moduleMap[name];
    const planPath = path.join(this.workspacePath!, mod.path, 'sqitch.plan');
    writeFileSync(planPath, plan);
  }

}
