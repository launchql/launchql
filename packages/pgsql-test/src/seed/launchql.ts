import { SeedAdapter, SeedContext } from './types';
import { getEnvOptions } from '@launchql/types';
import { LaunchQLProject, deployFast } from '@launchql/core';

export function launchql(cwd?: string, cache: boolean = false): SeedAdapter {
  return {
    async seed(ctx: SeedContext) {
      const proj = new LaunchQLProject(cwd ?? ctx.connect.cwd);
      if (!proj.isInModule()) return;

      const opts = getEnvOptions({ pg: ctx.config });

        await deployFast({
          opts,
          name: proj.getModuleName(),
          database: ctx.config.database,
          dir: proj.modulePath,
          usePlan: true,
          cache
        });
    }
  };
}
