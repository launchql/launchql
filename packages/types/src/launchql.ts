import { PostGraphileOptions } from 'postgraphile';
import type { Plugin } from 'graphile-build';
import { execSync } from 'child_process';
declare module 'express-serve-static-core' {
    interface Request {
        apiInfo: {
            data: {
                api: {
                    dbname: string;
                    anonRole: string;
                    roleName: string;
                    schemaNames: {
                        nodes: { schemaName: string }[];
                    };
                    schemaNamesFromExt: {
                        nodes: { schemaName: string }[];
                    };
                    apiModules: {
                        nodes: {
                            name: string;
                            data?: any;
                        }[];
                    };
                    rlsModule?: {
                        authenticate?: string;
                        authenticateStrict?: string;
                        privateSchema: {
                            schemaName: string;
                        };
                    };
                };
            };
        };
        svc_key: string;
        clientIp?: string;
        databaseId?: string;
        token?: {
            id: string;
            user_id: string;
            [key: string]: any;
        };
    }
}

export interface PgConfig {
    host: string;
    port: number;
    user: string;
    password: string;
    database: string;
}

export interface PgTestConnectionOptions {
    rootDb?: string;
    template?: string;
    prefix?: string;
    extensions?: string[];
    cwd?: string;
    connection?: {
      user?: string;
      password?: string;
      role?: string;
    }
  }

export interface LaunchQLOptions {
    db?: Partial<PgTestConnectionOptions>;
    pg?: Partial<PgConfig>;
    graphile?: {
        isPublic?: boolean;
        schema?: string | string[];
        metaSchemas?: string[];
        appendPlugins?: Plugin[];
        graphileBuildOptions?: PostGraphileOptions['graphileBuildOptions'];
        overrideSettings?: Partial<PostGraphileOptions>;
        // Only used when useMetaApi is false
        anonRole?: string;
        // Only used when useMetaApi is false
        roleName?: string;
    };
    server?: {
        host?: string;
        port?: number;
        trustProxy?: boolean;
        origin?: string;
        strictAuth?: boolean;
        middleware?: {
            useMetaApi?: boolean;
            useAuth?: boolean;
            useCors?: boolean;
            useGraphile?: boolean;
            useFlush?: boolean;
            customMiddleware?: any[];
        };
    };
    features?: {
        simpleInflection?: boolean;
        oppositeBaseNames?: boolean;
        postgis?: boolean;
    };
    cdn?: {
        bucketName?: string;
        awsRegion?: string;
        awsAccessKey?: string;
        awsSecretKey?: string;
        minioEndpoint?: string;
    };
}


export const launchqlDefaults: LaunchQLOptions = {
    db: {
        rootDb: 'postgres',
        prefix: 'db-',
        extensions: [],
        cwd: process.cwd(),
        connection: {
            user: 'app_user',
            password: 'app_password',
            role: 'anonymous'
        }
    },
    pg: {
        host: 'localhost',
        port: 5432,
        user: 'postgres',
        password: 'password',
        database: 'postgres',
    },
    // TODO these should all be subset of postgraphile options since we do spread these when we call getSettings() in server
    graphile: {
        isPublic: true,
        schema: [],
        // meta has a ref to databaseId, hence the connection here...
        metaSchemas: ['collections_public', 'meta_public'],
        appendPlugins: [],
        overrideSettings: {},
        graphileBuildOptions: {},
        // Only used when useMetaApi is false
        anonRole: 'anonymous',
        // Only used when useMetaApi is false
        roleName: 'authenticated',        
    },
    server: {
        host: 'localhost',
        port: 3000,
        trustProxy: false,
        strictAuth: false,
        middleware: {
            useMetaApi: true,
            useAuth: true,
            useCors: true,
            useGraphile: true,
            useFlush: true,
            customMiddleware: []
        }
    },
    features: {
        simpleInflection: true,
        oppositeBaseNames: true,
        postgis: true,
    },
    cdn: {
        bucketName: 'test-bucket',
        awsRegion: 'us-east-1',
        awsAccessKey: 'minioadmin',
        awsSecretKey: 'minioadmin'
    }
};

export function getGitConfigInfo(): { username: string; email: string } {
  const isTestEnv =
    process.env.NODE_ENV === 'test' ||
    process.env.NODE_ENV === 'testing' || // fallback
    process.env.GITHUB_ACTIONS === 'true'; // GitHub Actions

  if (isTestEnv) {
    return {
      username: 'CI Test User',
      email: 'ci@example.com'
    };
  }

  try {
    const username = execSync('git config --global user.name', {
      encoding: 'utf8'
    }).trim();
    const email = execSync('git config --global user.email', {
      encoding: 'utf8'
    }).trim();

    return { username, email };
  } catch {
    throw new Error('Failed to retrieve global git config. Ensure git is configured.');
  }
}
