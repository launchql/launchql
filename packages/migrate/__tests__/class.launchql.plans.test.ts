import path from 'path';
import { LaunchQLProject } from '../src/class/launchql';

const fixture = (p: string[]) =>
  path.resolve(__dirname, '../../..', '__fixtures__', 'sqitch', ...p);

/**
 * Cleans up text by trimming lines, removing empty lines, and joining them.
 */
const cleanText = (text: string): string =>
  text
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .join('\n');

const getModuleProject = async (name: string): Promise<LaunchQLProject> => {
  const workspace = new LaunchQLProject(fixture(['launchql']));
  await workspace.init();

  const moduleMap = workspace.getModuleMap();
  const meta = moduleMap[name];
  if (!meta) throw new Error(`Module ${name} not found in workspace`);

  const modPath = fixture(['launchql', meta.path]);
  const mod = new LaunchQLProject(modPath);
  await mod.init();
  return mod;
};

describe('sqitch modules', () => {
  it('should be able to create a plan', async () => {
    const mod = await getModuleProject('totp');
    const plan = await mod.generateModulePlan({ projects: false });
    expect(cleanText(plan)).toEqual(
      cleanText(`
%syntax-version=1.0.0
%project=totp
%uri=totp
procedures/generate_secret 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/generate_secret`)
    );
  });

  it('should be able to create a plan with cross-project requires already in', async () => {
    const mod = await getModuleProject('utils');
    const plan = await mod.generateModulePlan({ projects: false });
    expect(cleanText(plan)).toEqual(
      cleanText(`
%syntax-version=1.0.0
%project=utils
%uri=utils
projects/totp/procedures/generate_secret [totp:procedures/generate_secret] 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add projects/totp/procedures/generate_secret
procedures/myfunction 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/myfunction`)
    );
  });

  it('should create a plan without options for projects and lose dependencies', async () => {
    const mod = await getModuleProject('secrets');
    const plan = await mod.generateModulePlan({ projects: false });
    expect(cleanText(plan)).toEqual(
      cleanText(`
%syntax-version=1.0.0
%project=secrets
%uri=secrets
procedures/secretfunction 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/secretfunction
`)
    );
  });

  it('should create a plan that references projects', async () => {
    const mod = await getModuleProject('secrets');
    const plan = await mod.generateModulePlan({ projects: true });
    expect(cleanText(plan)).toEqual(
      cleanText(`
%syntax-version=1.0.0
%project=secrets
%uri=secrets
procedures/secretfunction [totp:procedures/generate_secret pg-verify:procedures/verify_view] 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/secretfunction
`)
    );
  });

  it('can create a module plan using workspace.generateModulePlan()', async () => {
    const mod = await getModuleProject('totp');
    const plan = await mod.generateModulePlan({ projects: false });
  
    expect(cleanText(plan)).toContain('%project=totp');
    expect(cleanText(plan)).toContain('procedures/generate_secret');
    expect(cleanText(plan)).toMatchSnapshot();
  });

  it('can create a module plan using workspace.generateModulePlan() w/projects', async () => {
    const mod = await getModuleProject('totp');
    const plan = await mod.generateModulePlan({ projects: true });
  
    expect(cleanText(plan)).toContain('%project=totp');
    expect(cleanText(plan)).toContain('procedures/generate_secret');
    expect(cleanText(plan)).toMatchSnapshot();
  });

  it('ensures plan includes all resolved deploy steps', async () => {
    const mod = await getModuleProject('secrets');
    const plan = await mod.generateModulePlan({ projects: false });
    expect(plan).toMatch(/add procedures\/secretfunction/);
    expect(cleanText(plan)).toMatchSnapshot();
  });
  
  it('ensures plan includes all resolved deploy steps w/projects', async () => {
    const mod = await getModuleProject('secrets');
    const plan = await mod.generateModulePlan({ projects: true });
    expect(plan).toMatch(/add procedures\/secretfunction/);
    expect(cleanText(plan)).toMatchSnapshot();
  });
  
  
});
