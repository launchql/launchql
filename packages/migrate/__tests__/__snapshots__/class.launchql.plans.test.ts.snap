// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqitch modules can create a module plan using workspace.generateModulePlan() 1`] = `
"%syntax-version=1.0.0
%project=totp
%uri=totp
procedures/generate_secret 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/generate_secret"
`;

exports[`sqitch modules can create a module plan using workspace.generateModulePlan() w/projects 1`] = `
"%syntax-version=1.0.0
%project=totp
%uri=totp
procedures/generate_secret [pg-verify:procedures/verify_view] 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/generate_secret"
`;

exports[`sqitch modules ensures plan includes all resolved deploy steps 1`] = `
"%syntax-version=1.0.0
%project=secrets
%uri=secrets
procedures/secretfunction 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/secretfunction"
`;

exports[`sqitch modules ensures plan includes all resolved deploy steps w/projects 1`] = `
"%syntax-version=1.0.0
%project=secrets
%uri=secrets
procedures/secretfunction [totp:procedures/generate_secret pg-verify:procedures/verify_view] 2017-08-11T08:11:51Z launchql <launchql@5b0c196eeb62> # add procedures/secretfunction"
`;
